torch/distributed/_tensor/ops/tensor_ops.py

```diff
@@ -465,77 +508,102 @@ def cat_rule(op_schema: OpSchema) -> OutputSharding:
-    # to avoid _Partial in result.
+    # Make sure all tensors are replciated on cat dimension
     need_reshard = False
-    tensor_list_specs_after = []
+    tensor_list_specs_after: List[DTensorSpec] = []
     for spec in tensor_list_specs:
-        if is_tensor_dim_sharded(spec, dim=dim):
+        if not is_empty(spec) and (
+            is_tensor_dim_sharded(spec, dim=dim) or is_tensor_partial(spec)
+        ):
             need_reshard = True
             tensor_list_specs_after.append(
                 DTensorSpec(
                     mesh=spec.mesh,
-                    placements=unshard_tensor_dim(spec.placements, dim=dim),
+                    placements=replicate_tensor_dim(spec.placements, dim=dim),
                     tensor_meta=spec.tensor_meta,
                 )
             )
         else:
             tensor_list_specs_after.append(spec)
+
     tensor_list_specs = tensor_list_specs_after
 
-    # TODO: currently einop rule requires every character
-    # in result notation must have appeared in inputs
-    # so we temporarily design cat notation as
-    # "aij,bij->aij". Once we modify this requirement,
-    # we can switch to the more logically reasonable notation
-    # "aij,bij->cij"
-    alphabet = "abcdefghijklmnopqrstuvwxyz"
-    einop_notation_list = []
-
-    l = len(tensor_list_specs)
-    free_dim = alphabet[l : l + ndim - 1]
-    for i, spec in enumerate(tensor_list_specs):
-        if spec.ndim == ndim:
-            # rewrite concat dim
-            dim_word = free_dim[:dim] + alphabet[i] + free_dim[dim:]
-            einop_notation_list.append(dim_word)
+    # align non-cat dimensions placements based on reshard cost
+    non_empty_specs = [spec for spec in tensor_list_specs if not is_empty(spec)]
+    mesh = non_empty_specs[0].mesh
+    ndim = non_empty_specs[0].ndim
+    new_placements: List[Placement] = []
+    for mesh_dim in range(mesh.ndim):
+        # compute the minimum cost of resharding on this mesh_dim
+        if any(
+            spec.placements[mesh_dim] != non_empty_specs[0].placements[mesh_dim]
+            for spec in non_empty_specs
+        ):
+            # only reshard if there is a mismatch
+            need_reshard = True
+            reshard_cost = []
+            for shard_dim in range(ndim):
+                # compute the cost of resharding on this shard_dim
+                cost: float = 0.0
+                for spec in non_empty_specs:
+                    global_shape = spec.shape
+                    if global_shape[shard_dim] < mesh.size(mesh_dim):
+                        # found one tensor where the shard_dim is smaller than
+                        # mesh_dim. In this case, we cannot shard on this shard_dim,
+                        # and hence set cost to infinity.
+                        cost = +float("inf")
+                    elif (
+                        is_tensor_dim_sharded(spec, dim=shard_dim)
+                        or prod(global_shape) == 0
```

Comment:
`non_empty_specs` should not have `prod(global_shape) == 0` case?

Reply:
Unfortunately this is possible due to `torch.Size[0, 2]` considered as non-empty by `torch.cat`, and we actually have test case touching this combination.